### ISSUE TITLE
Implement reply methods n lua-tg.c

### ISSUE THREAD

**Issue by luksireiku** [vysheng/tg#715]
lua-tg.c has almost all methods of tg-cli but it lacks methods to reply to messages, I want to make a telegram-bot to reply to messages and I can't, yet.

**Comment by drunkwolfs** [comment]
Use python

**Comment by LukeLR** [comment]
I don't think that "use Python" is an appropriate answer when the feature doesn't exist in lua (yet), because someone might prefer lua over Python, maybe because someone is experienced in lua but not in Python, and therefore I think we should keep both options on the same level, especially because lua was first and shouldn't be replaced by Python. So maybe someone who understands the lua scripting plugin might implement a method for replying.

**Comment by LukeLR** [comment]
To solve your problem, an alternative way of replying on messages exists:
In your `on_msg_receive(msg)` -method the msg-array is passed, which contains the sender of the original message. If you want your bot to reply on that message, you will just need to send a message to the sender of the incoming message. A possible way might look like this: `send_msg(msg.from, <text here>)` The original message won't be quoted in the reply as it would be when the real reply method was used, but this is the only disadvantage, if not even an advantage, because if every message was quoted in the chat history, it would appear twice, which, at least in my eyes, might be annoying ;) So maybe give it a try ;)

**Comment by luksireiku** [comment]
Thanks for trying to help, but the problem is that I need the quoted original message, because my script uses some data from the original message, that's why I need the reply method.

**Comment by LukeLR** [comment]
You don't need to reply to that message to get data from it! The content of the message is always available as `msg.text` :wink: So, in your `on_msg_receive` you can easily check the contents of the message, and, depending what content it has, do something specific and send a specific reply depending on the content. So, for example, if the user sends `ping` the script sends a reply with `pong` but sending `PING` will reply by forwarding the message the user sent back to himself. This is done in the default lua script, so check out test.lua in this repo, especially the on_msg_receive-method, and let me know if this doesn't solve the problem ;)

**Comment by luksireiku** [comment]
You don't got it, I need both the quoted message (msg.reply_to_message.text) and the reply (msg.text).

An example: BOT A sends '!file {url} {chat id}' to BOT B which downloads the file from the url and replies sending the file, then BOT A gets the file id and resends it to the chat id that gets from the original message.

**Comment by LukeLR** [comment]
Okay, I see, it's difficult because you need to transfer both an image and a text in the same message. An option would be to rename the file that was downloaded by the chat id, so BOT A downloads the file and forwards it to the chat id provided in the filename. You might tell BOT B to always send the chat id after returning the queried image, then BOT A could save the received image to a given filename, and when it receives a chat id, check if that file exists and then send it to the chat id. Every other option would include saving the chat id that the image should be sent to on BOT A, then send the request for downloading the file to BOT B, and when BOT A receives the file by BOT B, it reads the previously saved filename, and then sends the file to it. To make this support several image requests simultaneously, the file that the chat id is saved in should have the structure of a queue, e.g. a text file where the chat IDs of new requests are always appended to the end of the file, and when the requested image by BOT B arrives, the first chat id in the file is read and removed. The only difficulty could be if the order of the images that are incoming from BOT B isn't the same as the order of the requests, probably because some files take longer than others. To prevent this, you could use a) an external application, e.g. some application that contains the bot code and controls/reads from the telegram cli over a socket, or you could try Python with it's reply method providing both files and text in a single message.

One last question: why do you use two bots at all? Downloading the image and sending it to the requested chat id could be accomplished by a single bot…

**Comment by luksireiku** [comment]
1) BOT A uses bot api and it doesn't allow to download files.
2) Sending the chat id after the image means that the bot should listen to two messages and it may be a bit more complex to implement.
3) Anyway the easiest way to do this stuff is with a simple reply, I already tried it manualy using 'reply_photo {msg id} {photo}' in tg-cli and works like a charm, that's why I want to use the reply method too from lua.

Its just an example, the bots that use the new telegram bot api can't read messages from other api bots, so I use a tg-cli bot to communicate between api bots, sending images it's just an example, anyway if BOT A receives too many photo requests it doesn't became slow or crashes because BOT B downloads and uploads the photos without slowing BOT A, then the first bot resends the photo getting the file_id from the photo that BOT B sent.

**Comment by LukeLR** [comment]
Ah okay, Using the bot API of course Limits the functionalities of bot A. Then you really don't have the choice of avoiding the reply function. you probably should take a deeper look into Python or into the tg-cli lua interface…

**Comment by nullifye** [comment]
reply method succesfully implemented in Lua.
i hv made a pull request https://github.com/vysheng/tg/pull/800
